src/net.cpp

```diff
@@ -1090,6 +1133,35 @@ static void AcceptConnection(const ListenSocket& hListenSocket) {
         }
     }
 
+    // BU - add inbound connection to the ip tracker and increment counter
+    // If connection attempts exceeded within allowable timeframe then ban peer
+    {
+        double nConnections = 0;
+        LOCK(cs_mapInboundConnectionTracker);
+        CNetAddr ipAddress = (CNetAddr)addr;
+        if (mapInboundConnectionTracker.count(ipAddress)) {
+            // Decay the current number of connections (over 60 seconds) depending on the last connection attempt
+            int64_t nTimeElapsed = GetTime() - mapInboundConnectionTracker[ipAddress].nLastConnectionTime;
+            double nRatioElapsed = (double)nTimeElapsed / 60;
+            nConnections = mapInboundConnectionTracker[ipAddress].nConnections - (nRatioElapsed * mapInboundConnectionTracker[ipAddress].nConnections);
+            if (nConnections < 0) 
+                nConnections = 0;
+        }
+
+        nConnections += 1;
+        mapInboundConnectionTracker[ipAddress].nConnections = nConnections;
+        mapInboundConnectionTracker[ipAddress].nLastConnectionTime = GetTime();
+
+        LogPrint("evict", "Number of Connection attempts is %f for %s\n", nConnections, addr.ToString());
+        if (nConnections > 4) {
+            int nHoursToBan = 4;
+            CNode::Ban((CNetAddr)addr, BanReasonNodeMisbehaving, nHoursToBan*60*60);
+            LogPrintf("Banning %s for %d hours: Too many connection attempts - connection dropped\n", addr.ToString(), nHoursToBan);
+            return;
```

Comment:
I was letting the ban take effect on the next connection that comes in
becuase we don't want to disconnect whitelisted peers, but perhaps the
code is not entirely clear.  I should add that in and then we also don't
want to print out the log statement if the peer is whitelisted..so I'll
make that change, thanks

On 02/09/2016 9:26 PM, Justaphf wrote:
>
> In src/net.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/62#discussion_r77431155>:
>
> > +            double nRatioElapsed = (double)nTimeElapsed / 60;
> > +            nConnections = mapInboundConnectionTracker[ipAddress].nConnections - (nRatioElapsed * mapInboundConnectionTracker[ipAddress].nConnections);
> > +            if (nConnections < 0) 
> > +                nConnections = 0;
> > +        }
> > +
> > +        nConnections += 1;
> > +        mapInboundConnectionTracker[ipAddress].nConnections = nConnections;
> > +        mapInboundConnectionTracker[ipAddress].nLastConnectionTime = GetTime();
> > +
> > +        LogPrint("evict", "Number of Connection attempts is %f for %s\n", nConnections, addr.ToString());
> > +        if (nConnections > 4) {
> > +            int nHoursToBan = 4;
> > +            CNode::Ban((CNetAddr)addr, BanReasonNodeMisbehaving, nHoursToBan*60*60);
> > +            LogPrintf("Banning %s for %d hours: Too many connection attempts - connection dropped\n", addr.ToString(), nHoursToBan);
> > +            return;
>
> I think you need to call |CloseSocket(hSocket)| here since the socket
> is open but because of the ban we are returning.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/62/files/2192d2c5b6cda21ed24bd4a5928d88893087ac71#r77431155>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0MG6oosPhkWw0MepteN05D5aGY5mks5qmPbagaJpZM4Jc23I>.
>